tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
+      continue;
+    dyn_sizes.push_back(
+        b.create<memref::DimOp>(loc, tensor_vec[arg_number], dim_ind_pos));
+  }
+  return dyn_sizes;
+}
+
+// Adds indices/axes that are missing from output set
+llvm::SmallSetVector<StringRef, 4> FindSummationAxes(
+    const llvm::SmallSetVector<StringRef, 4>& input_set,
+    const llvm::SmallSetVector<StringRef, 4>& output_set) {
+  llvm::SmallSetVector<StringRef, 4> summation_axes;
+  for (const StringRef& ind : input_set) {
+    if (!output_set.contains(ind)) summation_axes.insert(ind);
+  }
+  return summation_axes;
+}
+
+// Given a 1:1 map from std::string -> affine dimension expression
+// we can get the affine expression of dimensions that an
+// operand will access based on the input_str of einsum_config
+// For example:
+// let string_dim_umap = {'a' : d0, 'b' : d1, 'c' : d2}
+// for einsum_config "abc,cb->acb"
+// the first_input_operand will get umap[{"a","b","c"}] -> (d0, d1, d2)
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr> GetExprFromConfig(
+    const llvm::SmallSetVector<StringRef, 4>& input_set,
+    const SmallVector<std::string>& loop_dims,
+    const DenseMap<StringRef, AffineExpr>& str_affine_dim_umap) {
+  SmallVector<AffineExpr> output_exprs;
+  for (const StringRef& dim : loop_dims) {
+    output_exprs.push_back(str_affine_dim_umap.lookup(dim));
+  }
+  return output_exprs;
+}
+
+// Convert mhlo.einsum op into linalg.generic
+// Algorithm in general 3 steps:
+
+// Step1) Dissect entire einsum_config to different operands
+// e.g f("abc,cd->abd") = {lhs:["abc"], rhs:["cd"], out:["abd"]}
+
+// Step2) Split up the string into vector of the elements
+// e.g {lhs:["abc"], rhs:["cd"], out:["abd"]} = {lhs:["a","b","c"],
+// rhs:["c","d"], out:["a","b","d"]}
+
+// Step3) Convert the vector into data access
+// patern represented by affineMaps with affineDimensions e.g
+// {lhs:["a","b","c"], rhs:["c","d"], out:["a","b","d"]} = {lhs:[d0,d1,d2],
+// rhs:[d2,d3], out:[d0,d1,d3]}
+
+// Note: Step2 is especially important to for batch
+// dim representation using ellipsis, S.T it has no limits to number of
+// dimension that the ellipsis can represent
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config();
+
+    // Setting Up Commonly Used Symbols as strings
+    const std::string kArrow = "->";
+    const std::string comma = ",";
+    const std::string ellipsis = "...";
```

Comment:
style nit, see https://google.github.io/styleguide/cppguide.html#Constant_Names
```suggestion
    const std::string kComma = ",";
    const std::string kEllipsis = "...";
```

Reply:
done